tee_supplicant/src/tee_supplicant.c

```diff
@@ -0,0 +1,295 @@
+		 * of the buffer.
+		 */
+		return 0;
+	}
+
+	sz = MEMREF_SIZE(params + idx);
+	offs = MEMREF_SHM_OFFS(params + idx);
+	if ((sz + offs) < sz)
+		return -4;
+	if ((sz + offs) > tshm->size)
+		return -5;
+
+	shm->flags = TEEC_MEM_INPUT | TEEC_MEM_OUTPUT;
+	shm->size = sz;
+	shm->id = MEMREF_SHM_ID(params + idx);
+	shm->buffer = (uint8_t *)tshm->p + offs;
+
+	return 0;
+}
+
+static void uuid_from_octets(TEEC_UUID *d, const uint8_t s[TEE_UUID_LEN])
+{
+	d->timeLow = (s[0] << 24) | (s[1] << 16) | (s[2] << 8) | s[3];
+	d->timeMid = (s[4] << 8) | s[5];
+	d->timeHiAndVersion = (s[6] << 8) | s[7];
+	memcpy(d->clockSeqAndNode, s + 8, sizeof(d->clockSeqAndNode));
+}
+
+static int load_ta(uint32_t num_params, struct tee_param *params)
+{
+	int ta_found = 0;
+	size_t size = 0;
+	struct param_value *val_cmd = NULL;
+	TEEC_UUID uuid;
+	TEEC_SharedMemory shm_ta;
+
+	memset(&uuid, 0, sizeof(uuid));
+	memset(&shm_ta, 0, sizeof(shm_ta));
+
+	if (num_params != 2 || get_value(num_params, params, 0, &val_cmd) ||
+	    get_param(num_params, params, 1, &shm_ta))
+		return TEEC_ERROR_BAD_PARAMETERS;
+
+	uuid_from_octets(&uuid, (void *)val_cmd);
+
+	size = shm_ta.size;
+	ta_found = TEECI_LoadSecureModule(&uuid, shm_ta.buffer, &size);
+	if (ta_found != TA_BINARY_FOUND) {
+		LOG_ERR("  TA not found");
+		return TEEC_ERROR_ITEM_NOT_FOUND;
+	}
+
+	MEMREF_SIZE(params + 1) = size;
+
+	/*
+	 * If a buffer wasn't provided, just tell which size it should be.
+	 * If it was provided but isn't big enough, report an error.
+	 */
+	if (shm_ta.buffer && size > shm_ta.size)
+		return TEEC_ERROR_SHORT_BUFFER;
+
+	return TEEC_SUCCESS;
+}
+
+static int process_request(const struct device *dev)
+{
+	int rc;
+	union tee_supp_msg ts_msg = {
+		.req.num_param = TEE_REQ_PARAM_MAX,
+	};
+
+	rc = receive_request(dev, &ts_msg.req);
+	if (rc) {
+		return rc;
+	}
+
+	LOG_INF("Receive OPTEE request cmd #%d", ts_msg.req.cmd);
+	switch (ts_msg.req.cmd) {
+	case OPTEE_MSG_RPC_CMD_LOAD_TA:
+		rc = load_ta(ts_msg.req.num_param, ts_msg.req.params);
+		break;
+	default:
+		return TEEC_ERROR_NOT_SUPPORTED;
+	}
+
+	ts_msg.rsp.ret = rc;
+	return send_response(dev, &ts_msg.rsp);
+}
+
+static void tee_supp_main(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+	const struct device *dev = p1;
+	int rc = 0;
+
+	while (1) {
+		rc = process_request(dev);
+		if (rc) {
+			LOG_ERR("Failed to process request, rc = %d", rc);
+			break;
+		}
+	}
+}
+
+int tee_supp_init(const struct device *dev)
+{
+	const struct device *tee_dev = DEVICE_DT_GET_ONE(linaro_optee_tz);
+
+	if (!tee_dev) {
+		LOG_ERR("No TrustZone device found!");
+		return -ENODEV;
+	}
+
+	k_thread_create(&main_thread, main_stack, K_THREAD_STACK_SIZEOF(main_stack), tee_supp_main,
+			(void *) tee_dev, NULL, NULL, TEE_SUPP_THREAD_PRIO, 0, K_NO_WAIT);
+
+	LOG_INF("Started tee_supplicant thread");
+	return 0;
+}
+
+void tee_supp_mutex_lock(struct k_mutex *mu)
+{
+	int e = k_mutex_lock(mu, K_FOREVER);
+
+	if (e) {
```

Comment:
AFAIK, error is not possible here with `K_FOREVER`

Reply:
Ok